internal/controller/thanosquery_controller.go

```diff
@@ -18,45 +18,142 @@ package controller
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.0/pkg/reconcile
 func (r *ThanosQueryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
 
-	// TODO(user): your logic here
+	query := &monitoringthanosiov1alpha1.ThanosQuery{}
+	err := r.Get(ctx, req.NamespacedName, query)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("thanos query resource not found. ignoring since object may be deleted")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get ThanosQuery")
+		return ctrl.Result{}, err
+	}
+
+	err = r.syncResources(ctx, *query)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *ThanosQueryReconciler) syncResources(ctx context.Context, query monitoringthanosiov1alpha1.ThanosQuery) error {
+	logger := log.FromContext(ctx)
+	var objs []client.Object
+
+	objs = append(objs, r.buildQuerier(query)...)
+
+	var errCount int32
+	for _, obj := range objs {
+		if manifests.IsNamespacedResource(obj) {
+			obj.SetNamespace(query.Namespace)
+			if err := ctrl.SetControllerReference(&query, obj, r.Scheme); err != nil {
+				logger.Error(err, "failed to set controller owner reference to resource")
+				errCount++
+				continue
+			}
+		}
+
+		desired := obj.DeepCopyObject().(client.Object)
+		mutateFn := manifests.MutateFuncFor(obj, desired)
+
+		op, err := ctrl.CreateOrUpdate(ctx, r.Client, obj, mutateFn)
+		if err != nil {
+			logger.Error(
+				err, "failed to create or update resource",
+				"gvk", obj.GetObjectKind().GroupVersionKind().String(),
+				"resource", obj.GetName(),
+				"namespace", obj.GetNamespace(),
+			)
+			errCount++
+			continue
+		}
+
+		logger.V(1).Info(
+			"resource configured",
+			"operation", op, "gvk", obj.GetObjectKind().GroupVersionKind().String(),
+			"resource", obj.GetName(), "namespace", obj.GetNamespace(),
+		)
+	}
+
+	if errCount > 0 {
+		return fmt.Errorf("failed to create or update %d resources for the hashrings", errCount)
+	}
+
+	return nil
+}
+
+func (r *ThanosQueryReconciler) buildQuerier(query monitoringthanosiov1alpha1.ThanosQuery) []client.Object {
+	metaOpts := manifests.Options{
+		Name:      query.GetName(),
+		Namespace: query.GetNamespace(),
+		Replicas:  query.Spec.Replicas,
+		Labels:    query.GetLabels(),
+		Image:     query.Spec.Image,
+		LogLevel:  query.Spec.LogLevel,
+		LogFormat: query.Spec.LogFormat,
+	}.ApplyDefaults()
+
+	return manifestquery.BuildQuerier(manifestquery.QuerierOptions{
+		Options:       metaOpts,
+		ReplicaLabels: query.Spec.QuerierReplicaLabels,
+		Timeout:       "15m",
+		LookbackDelta: "5m",
+		MaxConcurrent: 20,
+
+		// TODO(saswatamcode): Add logic to for automated tracking of endpoints.
```

Comment:
so I think what we will want to do here is add a watch to all Services (and maybe Endpoints/EndpointSlices to facilitate the other uses cases)

1. When we hit the watch event for a Service do a lookup of all ThanosQueries in the event handler
2. For each of the ThanosQueries, check if the Service labels matches one of the label selectors. 
3. If yes re-enqueue 
4. At reconcile time, lookup all the Services that match on the labels set in ThanosQuery
5. Convert to this type and pass to builder.

Reply:
Yup exactly what I was planning :)